### ISSUE TITLE
Memory leak in StrickMode

### ISSUE THREAD

**Issue by Aspsine** [square/retrofit#801]
04-03 15:48:22.135    9007-9017/com.aspsine.zhihu.daily E/StrictMode﹕ A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
    java.lang.Throwable: Explicit termination method 'close' not called
            at dalvik.system.CloseGuard.open(CloseGuard.java:184)
            at java.io.FileInputStream.<init>(FileInputStream.java:80)
            at okio.Okio.source(Okio.java:161)
            at com.squareup.okhttp.internal.io.FileSystem$1.source(FileSystem.java:44)
            at com.squareup.okhttp.internal.DiskLruCache$Entry.snapshot(DiskLruCache.java:1002)
            at com.squareup.okhttp.internal.DiskLruCache.get(DiskLruCache.java:430)
            at com.squareup.okhttp.Cache.get(Cache.java:180)
            at com.squareup.okhttp.Cache$1.get(Cache.java:139)
            at com.squareup.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:222)
            at com.squareup.okhttp.Call.getResponse(Call.java:271)
            at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:228)
            at com.squareup.okhttp.Call.getResponseWithInterceptorChain(Call.java:199)
            at com.squareup.okhttp.Call.execute(Call.java:79)
            at retrofit.client.OkClient.execute(OkClient.java:53)
            at retrofit.RestAdapter$RestHandler.invokeRequest(RestAdapter.java:326)
            at retrofit.RestAdapter$RestHandler.access$100(RestAdapter.java:220)
            at retrofit.RestAdapter$RestHandler$2.obtainResponse(RestAdapter.java:278)
            at retrofit.CallbackRunnable.run(CallbackRunnable.java:42)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at retrofit.Platform$Android$2$1.run(Platform.java:142)
            at java.lang.Thread.run(Thread.java:841)

**Comment by JakeWharton** [comment]
What version of Retrofit? What version of OkHttp? What does your endpoint declaration look like?

**Comment by Aspsine** [comment]
@JakeWharton 

Thanks.

### Version
compile 'com.squareup.retrofit:retrofit:1.9.0'
compile 'com.squareup.okhttp:okhttp:2.3.0'

### Code

I create a singleton of RestAdapter, OKHttpClient and DailyService.

 ```Java
public class RestApi {

    private static RestAdapter restAdapter;

    public static <T> T createApi(Class<T> clazz, String api) {
        if (restAdapter == null) {
            synchronized (RestApi.class) {
                if (restAdapter == null) {
                    restAdapter = new RestAdapter.Builder()
                            .setEndpoint(api)
                            .setClient(new OkClient(OkHttp.createHttpClient(App.getContext())))
                            .setConverter(new GsonConverter(new GsonBuilder().excludeFieldsWithoutExposeAnnotation().create()))
                            .setLogLevel(BuildConfig.DEBUG ? RestAdapter.LogLevel.FULL : RestAdapter.LogLevel.NONE)
                            .build();
                }
            }
        }
        return restAdapter.create(clazz);
    }
}


public class OkHttp {
    private static OkHttpClient okHttpClient;

    public static OkHttpClient createHttpClient(Context context) {
        if (okHttpClient == null) {
            synchronized (OkHttp.class) {
                okHttpClient = new OkHttpClient();
                okHttpClient.setCache(new Cache(FileUtils.getHttpCacheDir(context), Constants.Config.HTTP_CACHE_SIZE));
                okHttpClient.setConnectTimeout(Constants.Config.HTTP_CONNECT_TIMEOUT, TimeUnit.MILLISECONDS);
                okHttpClient.setReadTimeout(Constants.Config.HTTP_READ_TIMEOUT, TimeUnit.MILLISECONDS);
                okHttpClient.networkInterceptors().add(new CacheInterceptor());
            }
        }
        return okHttpClient;
    }
}


public class DailyApi {

    private static DailyApiService dailyApiService;

    public static DailyApiService createApi() {
        if (dailyApiService == null) {
            synchronized (DailyApi.class) {
                if (dailyApiService == null) {
                    dailyApiService = RestApi.createApi(DailyApiService.class, API);
                }
            }
        }
        return dailyApiService;
    }
}

private void refresh() {

        DailyApi.createApi().getTheme(mThemeId, new Callback<Theme>() {
            @Override
            public void success(Theme theme, Response response) {
                ...
            }

            @Override
            public void failure(RetrofitError error) {
                ...
            }
        });
}
```Java

**Comment by sohayb** [comment]
:+1:

**Comment by DannyHsuCC** [comment]
Any update for this memory leak ?

**Comment by JakeWharton** [comment]
This was fixed by #845 by adding `close()` to the exception catching response body which forwards to the delegate (the underlying stream).
